### PR TITLE
Redirect if the back end is called via the front end preview entry point

### LINKED ISSUE #3099: preview.php allows to create xml files with wrong urls.
**Affected version(s)**

Contao 4.9.15 and Contao 4.11.4

**Description**

The frontend preview allows you to call the backend via preview.php. For changes that lead to the generation of the XML files, the URLs are generated as preview.php URLs.

This happens easily when using the extension Frontend-Helper from Rocksolid. See also [https://github.com/madeyourday/contao-rocksolid-frontend-helper/issues/45](url)

Without the extension the behavior can be adjusted by: domain.tld/preview.php/contao?do=maintenance -> Rewrite XML files.

### PR DESCRIPTION
**Pull request by leofeyer**
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #3099
| Docs PR or issue | -

### CODE REVIEW

**Review by ausi — DISMISSED** [review]

**Review by bytehead — DISMISSED** [review]

**Comment by leofeyer** [comment]
What I dislike about the PR, is that the redirect only happens for the `contao_backend` route and not for the other `contao_*` routes such as `contao_backend_login` or `contao_backend_password`. @contao/developers WDYT?

**Comment by Toflar** [comment]
Imho the redirect should happen based on the scope? If it's not `frontend` then redirect?

**Review by aschempp — CHANGES_REQUESTED** [review]
I agree with @Toflar. I would only redirect backend scope though.

**Comment by leofeyer** [comment]
New concept added in fa97955ef8ad69f96267aa3882648b25c96e73ce. I will add the tests if we agree to go this way.

File: core-bundle/src/EventListener/BackendPreviewRedirectListener.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        if (!$event->isMasterRequest()) {
+            return;
+        }
+
+        $request = $event->getRequest();
+
+        if (!$request->attributes->has('_preview')
+            || true !== $request->attributes->get('_preview')
+            || !$this->scopeMatcher->isBackendRequest($request)
+            || ($request->attributes->has('_allow_preview') && true === $request->attributes->has('_allow_preview'))
```
Comment (ausi) [inline core-bundle/src/EventListener/BackendPreviewRedirectListener.php]:
```suggestion
            || ($request->attributes->has('_allow_preview') && true === $request->attributes->get('_allow_preview'))
```

File: core-bundle/src/EventListener/BackendPreviewRedirectListener.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        if (!$event->isMasterRequest()) {
+            return;
+        }
+
+        $request = $event->getRequest();
+
+        if (!$request->attributes->has('_preview')
+            || true !== $request->attributes->get('_preview')
+            || !$this->scopeMatcher->isBackendRequest($request)
```
Comment (ausi) [inline core-bundle/src/EventListener/BackendPreviewRedirectListener.php]: Would it make sense to check for the frontend instead?
```suggestion
            || $this->scopeMatcher->isFrontendRequest($request)
```

**Comment by ausi** [comment]
> New concept added in [fa97955](https://github.com/contao/contao/commit/fa97955ef8ad69f96267aa3882648b25c96e73ce).

I’m not 100% sure if I like the new `_allow_preview`, on the other hand hardcoding the two routes would make it inflexible for extension developers...

The rest LGTM! 🎉

File: core-bundle/src/EventListener/BackendPreviewRedirectListener.php (hunk 2 above)
Comment (leofeyer) [reply core-bundle/src/EventListener/BackendPreviewRedirectListener.php]: You mean in case someone has added their own scope?

**Comment by leofeyer** [comment]
> I’m not 100% sure if I like the new `_allow_preview`, on the other hand hardcoding the two routes would make it inflexible for extension developers...

Same.

**Comment by Toflar** [comment]
I like the concept, I have to admit 🤔

File: core-bundle/src/EventListener/BackendPreviewRedirectListener.php (hunk 2 above)
Comment (ausi) [reply core-bundle/src/EventListener/BackendPreviewRedirectListener.php]: Yes, or their own route without any scope.

Comment (ausi) [reply core-bundle/src/EventListener/BackendPreviewRedirectListener.php]: Idea © by @Toflar https://github.com/contao/contao/pull/3464#issuecomment-918095113

Comment (leofeyer) [reply core-bundle/src/EventListener/BackendPreviewRedirectListener.php]: Changed in cc58b21e612a18924c370a9cbc4ab83d68fb5278.

**Review by ausi — APPROVED** [review]

**Pull request merged**